### DISCUSSION TITLE
[WS2021] Homework 5 Discussion

### DISCUSSION THREAD

**Discussion by changkun** [mimuc/gp-ws2021#5] [Discussion] [answered]
This thread is open for the discussion of homework 5.

**Comment by torantie** [comment]
Hi since I am again stuck after a pretty decent time investment i thought its time for a few questions again. :D Since i already struggled in the lecture with the QEM it's probably not surprising that I am still a little confused.^^

If i understood correctly the first part to calculate the quadric error and the position of the target vector was to calculate the vertex quadrics. I assumed this means I have to sum the vertex quadric for every face attached to a vertex. 
The lecture slides description of the algorithm explains that next the edge quadrics should be calculated and stored in a priority queue. 
![image](https://user-images.githubusercontent.com/14915789/105880401-9f607800-6003-11eb-81c3-89a972fc3c7b.png)
I honestly do not know what "edge quadrics" means but i assumed that this is the quadric error. Based on the slides the quadric error can be calculated when i solve the linear equation:
![image](https://user-images.githubusercontent.com/14915789/105881220-860bfb80-6004-11eb-94e3-eb99e20423b1.png)
So assuming that i understood everything correctly thus far, the quadric error calculation is where i am stuck. This leaves me with several questions:
1. Is my assumption to calculate the vertex quadrics by summing Qi of every face attached to the vertex correct (see code below) ?
2. Is "edge quadrics" == quadric error ?
3. How can i solve the linear equation programmatically ? Hints would be appreciated.^^
```
  simplify(reduceRatio) {
    // TODO: implement the QEM simplification
    const pq = new PriorityQueue( (a, b) => a.halfedge.quadricError > b.halfedge.quadricError)

    let vertexQuadricsSum = new Array(this.vertices.length)

    for(let i = 0; i < this.vertices.length; i++){
      let vertex = this.vertices[i]

      //calculate the sum of all neighbouring faces
      let QSum = new Matrix()
      vertex.faces((face) => {
        QSum.add(this.calculateQi(vertex, face))
      })
      vertexQuadricsSum[vertex.idx] = QSum
    }

    for(let j = 0; j < this.vertices.length; j++){
      let vertex = this.vertices[j]
      
      //compute edge quadrics?
      vertex.halfedges((halfedge) =>{
        let nextVertex = halfedge.next.vertex
        halfedge.quadricError = vertexQuadricsSum[vertex.idx] + vertexQuadricsSum[nextVertex.idx]
      })
    }

    //TODO
    this.halfedges.forEach(e => pq.enqueue(e))

    for(let k = 0; k < reduceRatio; k++){
      const halfedge = pq.dequeue()

    }
  }


  calculateQi(vertex, face) {
    //unit face normal
    let uFN = face.normal().unit()
    let negatedUnitFaceNormal = uFN.scale(-1)

    //h = -ni^T * xi. Last Dimension for the homogeneous normal vector (x,y,z,h)
    let h = vertex.position.x * negatedUnitFaceNormal.x
        + vertex.position.y * negatedUnitFaceNormal.y
        + vertex.position.z * negatedUnitFaceNormal.z
    //Qi=ni*ni^T vertex Quadrics
    let Qi = new Matrix(
        uFN.x * uFN * x, uFN.x * uFN.y, uFN.x * uFN.z, uFN.x * h,
        uFN.y * uFN * x, uFN.y * uFN.y, uFN.y * uFN.z, uFN.y * h,
        uFN.z * uFN * x, uFN.z * uFN.y, uFN.z * uFN.z, uFN.z * h,
        h * uFN * x, h * uFN.y, h * uFN.z, h * h)
    return Qi;
  }
```

**Reply by changkun** [reply]
Hi, sorry for the very late response. It's been a busy week. I hope you already made few further thoughts ahead from solving the problem. Answering your questions:

> the first part to calculate the quadric error and the target vector's position was to calculate the vertex quadrics.

This is correct, calculating all vertex quadrics (4x4 matrices) is the first step of all computing. An edge quadric is the sum of its connected vertex quadrics as described in edge error:

```
class Edge {
  quadric() {
    const v1q = this.halfedge.vertex.quadric()
    const v2q = this.halfedge.next.vertex.quadric()
    return v1q.add(v2q) 
  }
  ...
```

> 1. Is my assumption to calculate the vertex quadrics by summing Qi of every face attached to the vertex correct (see code below) ?

Yes, each face normal contributes to a Qi and a vertex quadric is the sum of all Qi:

```
class Vertex {
  quadric() {
    let qq = new Matrix()
    this.faces(f => {
      qq = qq.add(f.quadric())
    })
    return qq
  }
  ...
```

> 2. Is "edge quadrics" == quadric error ?

See the beginning, an edge quadric is the sum of two vertex quadrics (still in 4x4 matrix), whereas edge error is the edge quadric multiply the best target vertex position $v^TQ_ev$.

> 3. How can I solve the linear equation programmatically? 

You don't need to solve the equation programmatically. It is a quadric equation, where you learned analytical solution from high school. If you try to write it down, you will end up with $ax^2+bx+c=0$ where $a,b,c$ are constant, and $x$ is the best target position you are looking for. (well, you will need to deal with cases where the edge quadric matrix cannot be inverted, in this case, do a few greedy searches that find a local optimal is good enough, such as divide the edge into 16 different parts, trying to multiply that vertex with the edge quadric, record and return the one with minimum error. This is an edge case where you could do it differently.)

I hope the answer helps a bit for you to solve the implementation.

**Reply by torantie** [reply]
> Yes, each face normal contributes to a Qi and a vertex quadric is the sum of all Qi

Cool thank you.^^

> See the beginning, an edge quadric is the sum of two vertex quadrics (still in 4x4 matrix), whereas edge error is the edge quadric multiply the best target vertex position $v^TQ_ev$.

Ah okay thank you. I think a definition for edge quadrics in the slides would be good since only vertex quadrics were defined.

> It is a quadric equation, where you learned analytical solution from high school. If you try to write it down, you will end up with $ax^2+bx+c=0$ where $a,b,c$ are constant...

i can't really see how
![image](https://user-images.githubusercontent.com/14915789/105881220-860bfb80-6004-11eb-94e3-eb99e20423b1.png)
resolves to ax^2+bx+c=0. Considering that x is a 4-dimensional vector (x,y,z,1) and a,b,c constants it doesn't really make sense for the solution to be a number. Doesn't the formula resolve to something like this for example:
![image](https://user-images.githubusercontent.com/14915789/106288698-a88e5680-6248-11eb-80ae-573422ada7ee.png)
![image](https://user-images.githubusercontent.com/14915789/106288736-b512af00-6248-11eb-8b2b-19d4bec0857e.png)
Sorry if the question is to low level.

**Reply by changkun** [reply]
Ah. Now I see where your confusion comes from. We are solving an optimization problem here, when I say "resolve ax^2+bx+c=0" meaning that taking it as a quadric function and compute the partial derivative of x, y, and z, so you will end up with three equations. For instance, in your provided example, the x partial derivative of the equation results in `4x+5y+6z+7=0` which is linear. You can do that similarly too for the rest two equations. Eventually, you will have linear equations based on the elements of the summed quadric matrix $Q'v = b'$ (try to derive Q' and b' yourself, it will be very simple and elegant :)  then solving the equation is by inverting the $Q'$, and $v = Q'^{-1}b'$.

**Reply by changkun** [reply]
> I think a definition for edge quadrics in the slides would be good since only vertex quadrics were defined.

Oppos.. Looks like I didn't put that information on slides... Sorry and thanks for pointing to the problem..

**Reply by torantie** [reply]
I'll be honest i am still not completely sure how to implement this, but thank you a lot for the explanation. One more question. Should we use the math library to calculate the linear equations or should this be achieved with only the provided classes?

**Reply by changkun** [reply]
You don't need any additional library. Inverting a 4x4 matrix is already provided in the `mat.js`.

> I'll be honest i am still not completely sure how to implement this

Ok, here is one more hint:

```
class Edge {
  bestV() { // find the target vertex
    const q = this.quadric() // edge quadrics
    if (Math.abs(q.det()) > 1e-3) { // if the matrix is invertable
      // if you verify Q' and b' mentioned in above, then you will realize that Q' is the edge quadrics and b' is a zero vector.
      const v = q.inv().mulpos(new Vector()) 
      if (v.x !== NaN && v.y !== NaN && v.z != NaN) {
        return v
      }
    }

    // if not invertable, then greedy search directly on the edge
    const n = 16
    const a = this.halfedge.vertex.position
    const b = this.halfedge.next.vertex.position
    const d = b.sub(a)
    let beste = -1.0
    let bestv = new Vector()
    for (let i = 0; i <= n; i++) {
      const t = i / n
      const v = a.add(d.scale(t))
      const e = q.quadricError(v) // you need implement this yourself, which is calculating v^T q v :)
      if (beste < 0 || e < beste) {
        beste = e
        bestv = v
      }
    }
    return bestv
  }
```

**Reply by torantie** [reply]
That is a big hint xD. I honestly thought i needed more than the provided matrix operations.
Thank you for the help.

**Reply by changkun** [reply]
@torantie  maybe you could try to think and try to submit your solution with a different strategy that deals with the case where the edge quadric is not invertible. The provided hint is just one possible way (which not entirely efficient, situation out of the theory :).

**Reply by torantie** [reply]
Good to know thanks i will try to come up with a different solution.

**Reply by changkun** [reply]
> ... where you learned analytical solution from high school

@torantie It just came to my mind that this sentence might be perceived in a very negative way, which I wrote it like "you don't know about high school knowledge". If this is the way you feel it, I must apologize and I absolutely didn't mean it in that way. I should rephrase the sentence to something like "where you have enough background knowledge but need some insights, maybe think more about this and that.. " since my original intention was to express that the equation is not that difficult when you figured it out, as you already did in your submission :)

**Comment by torantie** [comment]
I have another question is the edge error always lower or equal to zero?

**Reply by changkun** [reply]
Negative edge error indicates wrong calculation. Quadric errors are always positive.

For the bunny model we've been using for the whole semester, edge errors are close to zero at `1e-5`~`1e-6` scale because the model is small.

**Reply by torantie** [reply]
It can't be negative? But there is no absolute function for the edge error.
![image](https://user-images.githubusercontent.com/14915789/106480162-5ef17600-64ab-11eb-97db-4264dcc1aa8b.png)
Since the target vector x can be negative is it not possible for the formula solution to be negative?

**Reply by changkun** [reply]
Quadric errors are by definition positive since they are representing distances. (well, could be zero)

**Reply by nico778** [reply]
I had a similar problem with zero values, I used the matrix add function without realizing that it returns a new matrix.

**Reply by torantie** [reply]
ah yeah i did that too^^. The problem i described here was more about the values being negative. Unfortunately i can't find a solution to that except using a absolut function on the solution. Can't find an error in my calculations.
getVertexQuadric(){
```
    //calculate the sum of all neighbouring faces
    let QSum = new Matrix()
    this.faces((face) => {
      let Qi = this.calculateQi(this, face)
      QSum = QSum.add(Qi)
    })

    return QSum
  }

  calculateQi(vertex, face) {
    //unit face normal
    let uFN = face.normal().unit()
    let negatedUnitFaceNormal = uFN.scale(-1)

    //h = -ni^T * xi. Last Dimension for the homogeneous normal vector (x,y,z,h)
    let h = vertex.position.x * negatedUnitFaceNormal.x
        + vertex.position.y * negatedUnitFaceNormal.y
        + vertex.position.z * negatedUnitFaceNormal.z
    //Qi=ni*ni^T vertex Quadrics
    let Qi = new Matrix(
        uFN.x * uFN.x, uFN.x * uFN.y, uFN.x * uFN.z, uFN.x * h,
        uFN.y * uFN.x, uFN.y * uFN.y, uFN.y * uFN.z, uFN.y * h,
        uFN.z * uFN.x, uFN.z * uFN.y, uFN.z * uFN.z, uFN.z * h,
        h * uFN.x, h * uFN.y, h * uFN.z, h * h)
    return Qi;
  }
  
setEdgeQuadric(){
    //compute edge quadric
    let vertexQuadric = this.halfedge.vertex.getVertexQuadric()
    let nextVertexQuadric = this.halfedge.next.vertex.getVertexQuadric()
    this.edgeQuadric = vertexQuadric.add(nextVertexQuadric)

  }


  getQuadricError(edgeQuadric, targetVertexPosition){
    //(Q1+Q2) * x
    let v = edgeQuadric.mulpos(targetVertexPosition)

    let edgeError = v.x * targetVertexPosition.x
        + v.y * targetVertexPosition.y
        + v.z * targetVertexPosition.z

    return Math.abs(edgeError)
  }
```

**Reply by changkun** [reply]
@torantie Just a quick theory: if you found negative values, are they close to zero? If so then it should be a numerical stability issue which you can simply set as zero.

**Reply by torantie** [reply]
they are close to 0 but that is like you mentioned probably because of the small modell

**Reply by changkun** [reply]
For the bunny model, most of the errors are in the range [`1e-5`, `1e-6`] (if I remember correctly) so [`- 1e-10` , `- 1e-20`] range are just zero.

**Reply by torantie** [reply]
one value is -0.43427315013730605 for example

**Reply by changkun** [reply]
> they are close to 0 but that is like you mentioned probably because of the small modell

@torantie I think you don't fully get the numerical issue. We mentioned several times in the session about implementing these projects: relativity is important. What I mean by close is that compared to those positive values, if negative happens, then it is to compare them and see if it is just floating-point stability or real ones.


> one value is -0.43427315013730605 for example

Then there must be an implementation error other than floating point issue.

**Reply by changkun** [reply]
> one value is -0.43427315013730605 for example

I just tested a simple cube from my implementation, the error is expected to 2.666666666. As always, don't work directly on large meshes, always start from simple ones...

**Reply by krasnor** [reply]
> I just tested a simple cube from my implementation, the error is expected to 2.666666666.

Is 2.66 the max error or should all vertices have this error?
Did the tested cube have all its sides closed?

If I use a fully closed cube (12 faces) then the error will always be 1.
When trying it with the cube.obj from Homework 4 (the one where one side is left open), 
the results seem to go in the right direction (not quite sure about the 0 error value, I'll have to take a closer look tomorrow):

![image](https://user-images.githubusercontent.com/10096384/107127815-f6463700-68b8-11eb-91f1-da7a738f8db4.png)

**Reply by changkun** [reply]
@krasnor I don't remember the detail of the 2.66 but I think I did the following for the quick test:
1. a single cube created from Blender, then applied a triangulation modifier
2. export the triangle mesh as obj file
3. set reduction ratio as 0.5 then all decimated edges have an error of 2.66

Since the edge error might be different after edge collapse depends on the order of decimation for those edges that have the same error, but your first collapse should have 2.66 for the cube and others might be different.

**Comment by krasnor** [comment]
I'm not sure if it is intentional, but in Matrix.add() x30 and x01 seem not to be calculated correctly? 

![image](https://user-images.githubusercontent.com/10096384/106671783-4b2f3800-65af-11eb-9fb8-38a89c42b825.png)

Shouldn't this be the following?
```
    mm.x30 = this.x30 + m.x30
    mm.x01 = this.x01 + m.x01 
```

After changing it, the results did not change much for me (seem to still have somewhere else an error)

**Reply by changkun** [reply]
Oppos. This is super embarrassing.. might cause by vim bindings... Sorry again

**Reply by krasnor** [reply]
No problem :)
Small typos can always happen

**Comment by changkun** [comment, marked as answer]
Close due to the end of the submission period.
